sw/device/lib/dif/dif_kmac.c

```diff
@@ -4,4 +4,478 @@
 
 #include "sw/device/lib/dif/dif_kmac.h"
 
-// TODO: implement!
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+
+#include "kmac_regs.h"
+
+enum {
+  /**
+   * The maximum amount of usable bits in the output state.
+   *
+   * This constant may be assumed to be a multiple of 32.
+   *
+   * The actual number of usable bits may be lower than the value defined
+   * depending on the mode in use. The intent is that this constant is useful
+   * for sizing fixed length buffers.
+   */
+  kDifKmacMaximumBitRate = 1600 - (2 * 128),
+
+  /**
+   * The offset of the second share within the output state register.
+   */
+  kDifKmacStateShareOffset = 0x100,
+};
+
+/**
+ * Report whether the hardware is currently idle.
+ *
+ * If the hardware is not idle then the `CFG` register is locked.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently idle or not.
+ */
+static bool is_state_idle(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_IDLE_BIT);
+}
+
+/**
+ * Report whether the hardware is currently in the absorb state and accepting
+ * writes to the message FIFO.
+ *
+ * Note that writes to the message FIFO may still block if it is full.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently absorbing or not.
+ */
+static bool is_state_absorb(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_ABSORB_BIT);
+}
+
+/**
+ * Report whether the hardware is currently in the squeeze state which means
+ * that the output state is valid and may be read by software.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently in the squeeze state or not.
+ */
+static bool is_state_squeeze(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_SQUEEZE_BIT);
+}
+
+/**
+ * Wrapper around mmio_region_write32 that reads back the register after
+ * it has been written to verify that it was written successfully.
+ *
+ * @param base_addr Base address.
+ * @param offset Offset of target register.
+ * @param value The value to write to the register.
+ * @returns Whether the register was successfully set to `value`.
+ */
+static bool checked_mmio_region_write32(mmio_region_t base_addr,
```

Comment:
Yeah I think it can be removed.

I originally added it to check that writes to the `CFG` register actually occurred since it `CFG` is locked when an operation is in progress. I wasn't sure whether we needed to be careful about races with another piece of hardware (i.e. key manager). I don't think that is an issue any more since we are going to enforce exclusive access in the software anyway.

Reply:
Done, removed.